### PR TITLE
Pass doc_lens to OLMo-core in forward_for_logprobs

### PR DESCRIPTION
**Pull request by finbarrtimbers**
Fix OLMo-core GRPO intra-document attention by passing packed-sequence `doc_lens` / `max_doc_lens` (computed from `attention_mask`) directly into the OLMo-core model forward in `grpo_utils.forward_for_logprobs` so we can correctly support packing.

### CODE REVIEW

File: open_instruct/grpo_utils.py [hunk 1]
```diff
@@ -498,6 +498,28 @@ def compute_grpo_loss(
...
+    rows: list[torch.Tensor] = []
+    for i in range(B):
+        vals, counts = torch.unique_consecutive(attention_mask[i], return_counts=True)
+        rows.append(counts[vals > 0])
+    max_docs = max((t.numel() for t in rows), default=0)
+    doc_lens = torch.zeros((B, max_docs), dtype=torch.int32, device=attention_mask.device)
+    for i, t in enumerate(rows):
+        if t.numel():
+            doc_lens[i, : t.numel()] = t.to(torch.int32)
+    max_doc_lens = [int(t.max().item()) if t.numel() else 0 for t in rows]
```
[the remark this answers is not in the document]
Comment (finbarrtimbers) [reply open_instruct/grpo_utils.py]: Already addressed in eed344e59 — `compute_olmo_core_doc_lens` now uses `max_doc_lens = doc_lens.max(dim=1).values.tolist()`.

File: open_instruct/olmo_core_train_modules.py [hunk 2]
```diff
@@ -458,6 +460,7 @@ def train_batch(self, batch: dict[str, Any], dry_run: bool = False) -> None:
                     self.pad_token_id,
                     self.temperature,
                     return_entropy=self.grpo_config.record_entropy,
+                    pass_olmo_core_doc_lens=True,
```
[the remark this answers is not in the document]
Comment (finbarrtimbers) [reply open_instruct/olmo_core_train_modules.py:481]: Addressed in b333c9267 — `GRPOTrainModule.__init__` now asserts `attn_implementation` is one of `flash_2 / flash_3 / flash_4`, so non-flash backends fail early at construction with a clear message instead of inside the attention forward.

File: open_instruct/grpo_utils.py [hunk 3]
```diff
@@ -498,6 +498,28 @@ def compute_grpo_loss(
...
+    The packed-sequence collator encodes per-doc indices as positive integers (1, 2, ...) in
+    ``attention_mask``, with 0 for padding. OLMo-core's Transformer requires explicit
+    ``doc_lens`` (B, max_docs) int32 and ``max_doc_lens`` (list[int], per row) to do intra-doc
+    attention; passing ``position_ids`` alone is a silent no-op.
+    """
+    B = attention_mask.size(0)
+    rows: list[torch.Tensor] = []
+    for i in range(B):
+        vals, counts = torch.unique_consecutive(attention_mask[i], return_counts=True)
+        rows.append(counts[vals > 0])
```
[the remark this answers is not in the document]
Comment (finbarrtimbers) [reply open_instruct/grpo_utils.py]: Already addressed — `compute_olmo_core_doc_lens` uses `torch.unique_consecutive(..., return_counts=True)` and keeps the zero-valued padding span as its own `doc_lens` entry, so the cumulative offsets OLMo-core builds across the flattened batch stay aligned with each row. This is documented in the function docstring (commit fd8a55fb1).

**Review by hamishivi — APPROVED** [review]
happy to merge if this is tested to run fine

**Comment by finbarrtimbers** [comment]
Yeah, I tested it and it runs fine (this was actually a key bug that made `grpo.py` deviate from `grpo_fast.py`).
